### PR TITLE
[JENKINS-63366] Apply credentials masking first

### PR DESCRIPTION
**Pull request by jvz**
This rearranges the ordering of the configured log filters for a step by
applying the newly created filter as the first log filter instead of
last.

See [JENKINS-63366](https://issues.jenkins-ci.org/browse/JENKINS-63366).

@jglick @Wadeck @daniel-beck @jeffret-b

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 1]
```diff
@@ -187,53 +193,95 @@ private void doStart() throws Exception {
...
+    }
+
+    /** @deprecated Use {@link MaskingDecorator} */
+    @Deprecated
+    private static final class Filter extends ConsoleLogFilter implements Serializable {
+
+        private static final long serialVersionUID = 1;
+
+        @Override public OutputStream decorateLogger(AbstractBuild _ignore, final OutputStream logger) throws IOException, InterruptedException {
+            throw new UnknownServiceException();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:283]: ??

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 2]
```diff
@@ -187,53 +193,95 @@ private void doStart() throws Exception {
...
+            return new SerializationProxy(Secret.fromString(pattern.pattern()));
+        }
+
         private Object readResolve() throws ObjectStreamException {
-            if (this.charsetName == null) {
-                this.charsetName = Charsets.UTF_8.name();
+            throw new InvalidObjectException("Must use proxy");
+        }
+
+        private static class SerializationProxy implements Serializable {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:217]: Seems like overkill. Just have `SecretPattern` compile its pattern on demand.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 1 above)
Comment (jvz) [reply src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:283]: Typo most likely; I must've hit tab after typing `throw new Un`. Should be `UnsupportedOperationException`.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:283]: I think you must retain compatibility here. If someone has a build running with credential masking active, upgrades the plugin, and restarts Jenkins, this method will be called after the restart.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 2 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:217]: You just have

```java
private synchronized Pattern pattern() {
  if (pattern == null) {
    pattern = Pattern.compile(secret);
  }
  return pattern;
}
```

so it is compiled at most once per actual instance of `SecretPattern`.

Comment (jvz) [reply src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:217]: Then I think this class is only here due to best practices with serialization (i.e., use a serialization proxy unless your class is a pure data POJO of some sort with no enforced invariants).

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:217]: Just suggesting a simplification that avoids the need for `writeReplace` / `readResolve` tricks.

**Pull request closed without merging**